Mathlib/Order/LawfulPreorder.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2017 Pierre Quinton. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pierre Quinton
+-/
+import Mathlib.Order.SetNotation
+import Mathlib.Order.Bounds.Defs
+import Mathlib.Order.Defs.PartialOrder
+
+/-!
+# Lawful Suprema and Infima
+
+This file introduces type classes that enforce correctness properties for suprema and infima
+operations on preorders, ensuring they return actual least upper bounds and greatest lower bounds
+whenever they exist.
+
+## Main declarations
+
+
+- `LawfulSupPreorder`: Typeclass ensuring that whenever a set `s` has a least upper bound, `sSup s`
```

Comment:
I preferred the name `LawfulSup`; it seems a bit redundant for the name to specify an order structure when we're already talking about suprema or infima.